File: src/backtracking/python/Solution131.py
class Solution:

    def partition(self, s: str) -> List[List[str]]:
        
        def is_palindrome(s, left, right):
            while left <right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -=1
            return True

        def backtrack(start, s, path, result):
        
            if start == len(s):
                result.append(path[:])
                return

            for end in range(start, len(s)):
                
                if is_palindrome(s, start, end):
                    path.append(s[start:end + 1])
                    backtrack(end + 1, s, path, result)
                    path.pop()
        
        result = []
        backtrack(0, s, [], result)

        return result

File: src/backtracking/Solution131.cpp
#include <vector>
using namespace std;

class Solution131
{
    bool is_palindrome(const string& s)
    {
        int l = 0;
        int r = s.size() - 1;
        while (l < r)
        {
            if (s[l] != s[r])
            {
                return false;
            }
            l++;
            r--;
        }

        return true;
    }

    void f(const string& s, vector<string>& partitions, vector<vector<string>>& result, int p)
    {
        if (p == s.size())
        {
            result.push_back(partitions);
            return;
        }

        for (int i = p; i < s.size(); ++i)
        {
            const auto sub_str = s.substr(p, i - p + 1);
            if (is_palindrome(sub_str))
            {
                partitions.push_back(sub_str);
                f(s, partitions, result, i + 1);
                partitions.pop_back();
            }
        }
    }

public:
    vector<vector<string>> partition(string s)
    {
        vector<vector<string>> result;
        vector<string> partitions;
        f(s, partitions, result, 0);
        return result;
    }
};
